Triggers.md

```diff
@@ -0,0 +1,234 @@
+
+### Can't the same be done with modules?
+Modules have the ability to build logic placed on top of keyspace events. However Modules provide a much more restrictive form of logic which is much harder to implement and dynamically replace. 
+By building this new logic inside the redis function library users will have a much more simple way to create stored procedures which are persistent, managed and replicated. 
+
+
+## Design and Specification
+
+### Triggers as part of function libraries
+Valkey triggers will be implemented as an extension of the Valkey [Function Libraries](https://valkey.io/topics/functions-intro/) .
+A trigger is basically a Valkey function which will be called when some specific keyspace event was issued.
+in order to register a specific function as trigger, one will simply have to use the new **"register_trigger"**  API which will be available for the different library engines. 
+For example for the (currently only) LUA engine the registration code will look like:
+
+![image](https://user-images.githubusercontent.com/88133677/225090299-1cf05508-75e6-43dd-a90f-05ca91dd3a6f.png)
+
+### Register trigger with named arguments form<
+ In order to provide some extra arguments to the trigger, the new API will also support the named arguments form. 
+ For example:
+
+![image](https://user-images.githubusercontent.com/88133677/225531522-d3f66a00-bd90-490b-98c8-11b62c3f3eba.png)
+
+### The Trigger code block
+As stated before, a trigger is basically a specific engine function.
+The function synopsis will be identical for all triggers which will get the following parameters:
+
+**a.     key -** the key for which the event was triggered
+
+**b.     event -** a string specifying the specific event (ie. “expire ”, “set”, “move”, “store” etc...)
+
+**c.     db -** the id of the db context for which the event is triggered (note it is not necessarily the current db scope the trigger is executed in, ie. MOVE event).
+
+For LUA engine based functions, The key will be supplied as the first argument in the keys array (eg KEYS[1]) 
+The **event** and **db** will be supplied in the ARGS array respectively (event will be provided as ARGS[1] and db as ARGS[2])
+
+
+### Triggering Events
+
+Valkey already has a feature to report different [keyspace notifications](https://valkey.io/topics/notifications/) which currently provides the ability to publish keyspace events to registered subscribers. The implementation should introduce the ability to set a trigger function to run on **ANY** event which is defined for [keyspace notifications](https://valkey.io/topics/notifications/).
+
+### Triggers configuration
+By design Triggers will not require ANY configuration in order to work. The Triggers will be part of the loadable function libraries and should be able to run on the target events even if the user **DID NOT** explicitly configured keyspace-notifications.
+
+### Atomicity and triggers
+
```

Comment:
>My understanding is that Valkey only offers a very limited form of atomicity. For simple commands, Valkey matches the general definition. MULTI-EXEC transactions can fail in the middle, so that changes prior to the failure happen, while changes after the failure do not. Can a trigger cause a MULTI-EXEC transaction to fail in the middle?

It is not just about "real fail" it is about logical failure. MULTI-EXEC can assume nothing changed once they were issued (using the key watch) however the key might still be changed via triggers while the transaction is executed. Maybe this is wanted and maybe not.

> It seems to me that in order to achieve any kind of atomicity, the trigger must be executed inside the same lock that guards the database update and scheduling of pubsub event notifications. Note: modules provide callbacks that will also be executed before this lock is released while pubsub event notifications are delivered after this lock is released.

So it is true that the module notifications are "delivered" once the action is performed. the keyspace notifications are also "written" at the same time, however the application receives them only later on and needs to handle the potential lagging in it's knowledge of the data state. In both options I described, the triggers are executed under the same "lock" meaning as part of the same "ExecutionUnit", the only question is if it should be done in the middle of the ExecutionUnit or at the end of it.

 IMO keeping the same way application reacted to keyspace operations without having to operate a complete late roundtrip following keyspace notification would greatly simplify the understanding of this feature, while also providing the ability to make sure it will be executed in the same ExecutionUnit.